pack.go

```diff
@@ -23,56 +23,108 @@ import (
-	Subject *artifactspec.Descriptor
-	// ManifestAnnotations is the annotation map of the manifest.
-	ManifestAnnotations map[string]string
+// Pack packs the given blobs, generates a manifest for the pack,
+// and pushes it to a content storage.
+// When opts.PackImageManifest is true, artifactType will be used as the
+// the config descriptor mediaType of the image manifest.
+// If succeeded, returns a descriptor of the manifest.
+func Pack(ctx context.Context, pusher content.Pusher, artifactType string, blobs []ocispec.Descriptor, opts PackOptions) (ocispec.Descriptor, error) {
+	if opts.PackImageManifest {
+		return packImage(ctx, pusher, artifactType, blobs, opts)
+	}
+	return packArtifact(ctx, pusher, artifactType, blobs, opts)
 }
 
-// Pack packs the given layers, generates a manifest for the pack,
-// and pushes it to a content storage.
+// packArtifact packs the given blobs, generates an artifact manifest for the
+// pack, and pushes it to a content storage.
+// If succeeded, returns a descriptor of the manifest.
+func packArtifact(ctx context.Context, pusher content.Pusher, artifactType string, blobs []ocispec.Descriptor, opts PackOptions) (ocispec.Descriptor, error) {
+	if artifactType == "" {
+		artifactType = MediaTypeUnknownArtifact
+	}
+
+	annotations, err := ensureAnnotationCreated(opts.ManifestAnnotations, ocispec.AnnotationArtifactCreated)
+	if err != nil {
+		return ocispec.Descriptor{}, err
+	}
+	manifest := ocispec.Artifact{
+		MediaType:    ocispec.MediaTypeArtifactManifest,
+		ArtifactType: artifactType,
+		Blobs:        blobs,
+		Subject:      opts.Subject,
+		Annotations:  annotations,
+	}
+	manifestJSON, err := json.Marshal(manifest)
+	if err != nil {
+		return ocispec.Descriptor{}, fmt.Errorf("failed to marshal manifest: %w", err)
+	}
+	manifestDesc := content.NewDescriptorFromBytes(ocispec.MediaTypeArtifactManifest, manifestJSON)
+	manifestDesc.ArtifactType = manifest.ArtifactType
+	manifestDesc.Annotations = manifest.Annotations
+
+	// push manifest
+	if err := pusher.Push(ctx, manifestDesc, bytes.NewReader(manifestJSON)); err != nil && !errors.Is(err, errdef.ErrAlreadyExists) {
+		return ocispec.Descriptor{}, fmt.Errorf("failed to push manifest: %w", err)
+	}
+
+	return manifestDesc, nil
+}
+
+// packImage packs the given blobs, generates an image manifest for the pack,
+// and pushes it to a content storage. artifactType will be used as the config
+// descriptor mediaType of the image manifest.
 // If succeeded, returns a descriptor of the manifest.
-func Pack(ctx context.Context, pusher content.Pusher, layers []ocispec.Descriptor, opts PackOptions) (ocispec.Descriptor, error) {
-	if opts.ConfigMediaType == "" {
-		opts.ConfigMediaType = MediaTypeUnknownConfig
+func packImage(ctx context.Context, pusher content.Pusher, configMediaType string, layers []ocispec.Descriptor, opts PackOptions) (ocispec.Descriptor, error) {
+	if configMediaType == "" {
+		configMediaType = MediaTypeUnknownConfig
```

Comment:
if `opts.PackImageManifest` is `true`, should we set this to `MediaTypeUnknownArtifact`?

Reply:
Sorry I don't get it 🤔

Reply:
Filled the gap offline. If oci artifact manifest is not supported, it's `Pack` callee's duty to repack a manifest and repush.